core/src/main/java/org/ehcache/EhcacheManager.java

```diff
@@ -334,21 +334,20 @@ public void close() {
     if(cacheLoaderWriterProvider != null) {
       loaderWriter = cacheLoaderWriterProvider.createCacheLoaderWriter(alias, config);
       WriteBehindConfiguration writeBehindConfiguration = ServiceLocator.findSingletonAmongst(WriteBehindConfiguration.class, config.getServiceConfigurations().toArray());
-      if(writeBehindConfiguration != null) {
-        final WriteBehindDecoratorLoaderWriterProvider factory = serviceLocator.getService(WriteBehindDecoratorLoaderWriterProvider.class);
-        decorator = factory.createWriteBehindDecoratorLoaderWriter((CacheLoaderWriter<K, V>)loaderWriter, writeBehindConfiguration);
+      if(writeBehindConfiguration == null) {
+        decorator = loaderWriter;
+      } else {
+        final WriteBehindProvider factory = serviceLocator.getService(WriteBehindProvider.class);
+        decorator = factory.createWriteBehindLoaderWriter(loaderWriter, writeBehindConfiguration);
         if(decorator != null) {
           lifeCycledList.add(new LifeCycledAdapter() {
             @Override
             public void close() {
-              factory.releaseWriteBehindDecoratorCacheLoaderWriter(decorator);
+              factory.releaseWriteBehindLoaderWriter(decorator);
```

Comment:
This cannot work with regards to lifecycle. The write behind stuff needs to be closed before the loader writer part, so the registration needs to be for loader writer first and write behind second.